### ISSUE TITLE
intermittent concurrent map writes - master

### ISSUE THREAD

**Issue by sstarcher** [roboll/helmfile#32]
fatal error: concurrent map writes

goroutine 6 [running]:
runtime.throw(0x11c4712, 0x15)
	/usr/local/Cellar/go/1.9.4/libexec/src/runtime/panic.go:605 +0x95 fp=0xc420043908 sp=0xc4200438e8 pc=0x10292f5
runtime.mapassign_faststr(0x1194640, 0xc420078510, 0x11c2e6b, 0xe, 0x12d3980)
	/usr/local/Cellar/go/1.9.4/libexec/src/runtime/hashmap_fast.go:861 +0x4da fp=0xc420043988 sp=0xc420043908 pc=0x100c5ba
text/template.(*Template).associate(0xc420056140, 0xc420056140, 0xc4200ce100, 0xe, 0x1, 0x2)
	/usr/local/Cellar/go/1.9.4/libexec/src/text/template/template.go:224 +0xad fp=0xc4200439c8 sp=0xc420043988 pc=0x10ff96d
text/template.(*Template).AddParseTree(0xc420056140, 0x11c2e6b, 0xe, 0xc4200ce100, 0x0, 0x0, 0x0)
	/usr/local/Cellar/go/1.9.4/libexec/src/text/template/template.go:128 +0x11c fp=0xc420043a30 sp=0xc4200439c8 pc=0x10fe8cc
text/template.(*Template).Parse(0xc420056140, 0xc420010340, 0x34, 0x10d1eaf, 0xc420048de0, 0x2)
	/usr/local/Cellar/go/1.9.4/libexec/src/text/template/template.go:205 +0x20a fp=0xc420043b40 sp=0xc420043a30 pc=0x10ff64a
github.com/roboll/helmfile/state.renderTemplateString(0xc420010340, 0x34, 0x2, 0xc420010340, 0x34, 0x0)
	/Users/sstarcher/go/src/github.com/roboll/helmfile/state/state.go:87 +0x45 fp=0xc420043b98 sp=0xc420043b40 pc=0x1148845
github.com/roboll/helmfile/state.flagsForChart(0xc42001a240, 0x22, 0xc420043f50, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/sstarcher/go/src/github.com/roboll/helmfile/state/state.go:246 +0x274 fp=0xc420043e10 sp=0xc420043b98 pc=0x11498f4
github.com/roboll/helmfile/state.(*HelmState).DiffCharts.func1(0xc420082500, 0xc42000adc0, 0x12d3218, 0x0, 0x0, 0x12a1820, 0xc420056980, 0xc4200165c0, 0xc420016410, 0xe, ...)
	/Users/sstarcher/go/src/github.com/roboll/helmfile/state/state.go:171 +0x71 fp=0xc420043f10 sp=0xc420043e10 pc=0x114b371
runtime.goexit()
	/usr/local/Cellar/go/1.9.4/libexec/src/runtime/asm_amd64.s:2337 +0x1 fp=0xc420043f18 sp=0xc420043f10 pc=0x1055841
created by github.com/roboll/helmfile/state.(*HelmState).DiffCharts
	/Users/sstarcher/go/src/github.com/roboll/helmfile/state/state.go:168 +0x1e4

goroutine 1 [semacquire]:
sync.runtime_Semacquire(0xc4200165cc)
	/usr/local/Cellar/go/1.9.4/libexec/src/runtime/sema.go:56 +0x39
sync.(*WaitGroup).Wait(0xc4200165c0)
	/usr/local/Cellar/go/1.9.4/libexec/src/sync/waitgroup.go:131 +0x72
github.com/roboll/helmfile/state.(*HelmState).DiffCharts(0xc420082500, 0x12a1820, 0xc420056980, 0x12d3218, 0x0, 0x0, 0x60000000008, 0x18, 0x18)
	/Users/sstarcher/go/src/github.com/roboll/helmfile/state/state.go:190 +0x239
main.main.func3(0xc42009a280, 0x0, 0x0)
	/Users/sstarcher/go/src/github.com/roboll/helmfile/main.go:150 +0x17f
reflect.Value.call(0x118c9c0, 0x11cc568, 0x13, 0x11c12b8, 0x4, 0xc420043778, 0x1, 0x1, 0xc4200436f8, 0x11b8e20, ...)
	/usr/local/Cellar/go/1.9.4/libexec/src/reflect/value.go:434 +0x905
reflect.Value.Call(0x118c9c0, 0x11cc568, 0x13, 0xc420043778, 0x1, 0x1, 0xc42009a280, 0x1319988, 0x0)
	/usr/local/Cellar/go/1.9.4/libexec/src/reflect/value.go:302 +0xa4
github.com/roboll/helmfile/vendor/github.com/urfave/cli.HandleAction(0x118c9c0, 0x11cc568, 0xc42009a280, 0x0, 0x0)
	/Users/sstarcher/go/src/github.com/roboll/helmfile/vendor/github.com/urfave/cli/app.go:483 +0x180
github.com/roboll/helmfile/vendor/github.com/urfave/cli.Command.Run(0x11c133c, 0x4, 0x0, 0x0, 0x0, 0x0, 0x0, 0x11caf4b, 0x33, 0x0, ...)
	/Users/sstarcher/go/src/github.com/roboll/helmfile/vendor/github.com/urfave/cli/command.go:186 +0x9ed
github.com/roboll/helmfile/vendor/github.com/urfave/cli.(*App).Run(0xc420001080, 0xc42000a060, 0x2, 0x2, 0x0, 0x0)
	/Users/sstarcher/go/src/github.com/roboll/helmfile/vendor/github.com/urfave/cli/app.go:237 +0x57d
main.main()
	/Users/sstarcher/go/src/github.com/roboll/helmfile/main.go:212 +0x9ea

goroutine 5 [runnable]:
text/template.(*Template).Execute(0xc420056140, 0x129f340, 0xc420100000, 0x0, 0x0, 0x0, 0x2)
	/usr/local/Cellar/go/1.9.4/libexec/src/text/template/exec.go:179 +0x76
github.com/roboll/helmfile/state.renderTemplateString(0xc4200d0000, 0x39, 0x2, 0xc4200d0000, 0x39, 0x0)
	/Users/sstarcher/go/src/github.com/roboll/helmfile/state/state.go:93 +0xad
github.com/roboll/helmfile/state.flagsForChart(0xc42001a240, 0x22, 0xc420077f50, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/sstarcher/go/src/github.com/roboll/helmfile/state/state.go:246 +0x274
github.com/roboll/helmfile/state.(*HelmState).DiffCharts.func1(0xc420082500, 0xc42000adc0, 0x12d3218, 0x0, 0x0, 0x12a1820, 0xc420056980, 0xc4200165c0, 0xc420018440, 0x13, ...)
	/Users/sstarcher/go/src/github.com/roboll/helmfile/state/state.go:171 +0x71
created by github.com/roboll/helmfile/state.(*HelmState).DiffCharts
	/Users/sstarcher/go/src/github.com/roboll/helmfile/state/state.go:168 +0x1e4

goroutine 7 [runnable]:
text/template/parse.(*lexer).emit(0xc4200ca180, 0x6)
	/usr/local/Cellar/go/1.9.4/libexec/src/text/template/parse/lex.go:156 +0xc5
text/template/parse.lexText(0xc4200ca180, 0x6)
	/usr/local/Cellar/go/1.9.4/libexec/src/text/template/parse/lex.go:269 +0x1e0
text/template/parse.(*lexer).run(0xc4200ca180)
	/usr/local/Cellar/go/1.9.4/libexec/src/text/template/parse/lex.go:231 +0x5f
created by text/template/parse.lex
	/usr/local/Cellar/go/1.9.4/libexec/src/text/template/parse/lex.go:224 +0x15f

goroutine 17 [runnable]:
runtime.goexit1()
	/usr/local/Cellar/go/1.9.4/libexec/src/runtime/proc.go:2365 +0x4e
runtime.goexit()
	/usr/local/Cellar/go/1.9.4/libexec/src/runtime/asm_amd64.s:2338 +0x6
created by text/template/parse.lex
	/usr/local/Cellar/go/1.9.4/libexec/src/text/template/parse/lex.go:224 +0x15f

**Comment by mumoshu** [comment]
Hi @sstarcher!
Out of curiosity, I've tried reproducing it myself, which failed.

Sounds like it is some kind of races like https://github.com/golang/go/issues/16101#issuecomment-253535084 but I can't say exactly without reproducing it myself.

How many value templates do you have in `set` fields of your charts.yaml?
Could you by any chance "enhance" your charts.yaml to trigger the race consistently? 😃

**Comment by mumoshu** [comment]
I suspect this part of the code-base to be the source of your problem:

https://github.com/roboll/helmfile/blob/c00b869045132fa636c8ee178d23b65dccf1a3cd/state/state.go#L72-L87

AFAIK, text/template is thread-safe only after constructed. 
So reusing `stringTemplate` across multiple `Parse` calls seems problematic.

**Comment by mumoshu** [comment]
Hi @fishi0x01!
Just wanted to reach out to you because you seem to be the original author of the feature introduced this usage of text/template 😄 

https://github.com/roboll/helmfile/pull/20/files#diff-f6be34387a141f652301c16cf7a5e36cR72

Do you have any clue on what's happening here?

**Comment by sstarcher** [comment]
I'm using a template {{}} about 10 times, and about 15 variable sets and 5 charts.  I have not been able to reproduce it easily it happens maybe 1/10 of the times I run it.

I have had similar issues reusing something like `template.New("stringTemplate")` in golang so your suspicion is reasonable.

**Comment by sstarcher** [comment]
I also built it with golang 1.9.4, but now that I think about it I don't know if I have reproduced it since I switched to the binary.

**Comment by mumoshu** [comment]
@sstarcher Thanks for the additional info!

I hope 1.9.4 fixes the problem. I don't see any relevant fix about that [in golang src](https://github.com/golang/go/blame/master/src/text/template/template.go), though.

I'll submit a PR for this soon. Would you mind trying that once you've reproduced the issue even in 1.9.4?

**Comment by sstarcher** [comment]
ya, testing it out now and trying to dig into the issue.  Found another issue while doing this.

**Comment by sstarcher** [comment]
just realized I was never using the built binary as the feature I'm using is only on master.  So I have always been using what was built with 1.9.4

**Comment by sstarcher** [comment]
This will reproduce it.  If you have a single chart the concurrency issue does not happen, but when you have 2 charts with environment variables that's when the concurrency jumps out.  I ran it 200 times with a single chart and no issues.  I ran it twice until it errored and it errored between 20-30 tries.

```
  - name: prometheus
    namespace: ops
    chart: stable/prometheus
    version: '5.1.5'
    set:
      - name: config.value
        value: '{{ env "HOME" }}'
      - name: config.value1
        value: '{{ env "HOME" }}'
      - name: config.value2
        value: '{{ env "HOME" }}'
      - name: config.value3
        value: '{{ env "HOME" }}'  
  - name: prometheus2
    namespace: ops
    chart: stable/prometheus
    version: '5.1.5'
    set:
      - name: config.value
        value: '{{ env "HOME" }}'
      - name: config.value1
        value: '{{ env "HOME" }}'
      - name: config.value2
        value: '{{ env "HOME" }}'
      - name: config.value3
        value: '{{ env "HOME" }}'    
```

**Comment by mumoshu** [comment]
Great! I've reproduced it myself with your example. Ran `for i in $(seq 1 100); do helmfile sync; done` and I got the error between 3rd and 5th run.

With the binary built from #38, I couldn't reproduce it anymore. How about you?

**Comment by sstarcher** [comment]
@mumoshu same your PR fixes my issue

**Comment by fishi0x01** [comment]
@mumoshu you are absolutely right. From the [docs](https://golang.org/pkg/text/template/): 

> Once parsed, a template may be executed safely in parallel, although if parallel executions share a Writer the output may be interleaved.

This indicates, that parallel `Parse` calls on the same template object are not allowed. Should have read the doc more carefully here .. Sorry for introducing this bug. 

Thx a lot for finding and fixing it!

**Comment by gtaylor** [comment]
Nice find! Thanks to both of you.

**Comment by roboll** [comment]
Thanks for chasing this @mumoshu :+1:
